source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,617 @@
+        USMAllocInfo{AllocBegin, UserBegin, UserEnd, NeededSize, Type});
+
+    // For updating shadow memory
+    {
+        std::scoped_lock<ur_shared_mutex> Guard(DeviceInfo->Mutex);
+        DeviceInfo->AllocInfos.emplace_back(AllocInfo);
+    }
+
+    // For memory release
+    {
+        std::scoped_lock<ur_shared_mutex> Guard(ContextInfo->Mutex);
+        ContextInfo->AllocatedUSMMap[AllocBegin] = AllocInfo;
+    }
+
+    context.logger.info(
+        "AllocInfos(AllocBegin={},  User={}-{}, NeededSize={}, Type={})",
+        (void *)AllocBegin, (void *)UserBegin, (void *)UserEnd, NeededSize,
+        Type);
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::releaseMemory(ur_context_handle_t Context,
+                                                void *Ptr) {
+    auto ContextInfo = getContextInfo(Context);
+
+    std::shared_lock<ur_shared_mutex> Guard(ContextInfo->Mutex);
+
+    auto Addr = (uptr)Ptr;
+    // Find the last element is not greater than key
+    auto AllocInfoIt = ContextInfo->AllocatedUSMMap.upper_bound((uptr)Addr);
+    if (AllocInfoIt == ContextInfo->AllocatedUSMMap.begin()) {
+        context.logger.error(
+            "Can't find release pointer({}) in AllocatedAddressesMap", Ptr);
+        return UR_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+    --AllocInfoIt;
+    auto &AllocInfo = AllocInfoIt->second;
+
+    context.logger.debug("USMAllocInfo(AllocBegin={}, UserBegin={})",
+                         AllocInfo->AllocBegin, AllocInfo->UserBegin);
+
+    if (Addr != AllocInfo->UserBegin) {
+        context.logger.error("Releasing pointer({}) is not match to {}", Ptr,
+                             AllocInfo->UserBegin);
+        return UR_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    // TODO: Update shadow memory
+    return context.urDdiTable.USM.pfnFree(Context,
+                                          (void *)AllocInfo->AllocBegin);
+}
+
+bool SanitizerInterceptor::preLaunchKernel(ur_kernel_handle_t Kernel,
+                                           ur_queue_handle_t Queue,
+                                           ur_event_handle_t &Event) {
+    prepareLaunch(Queue, Kernel);
+
+    UR_CALL(updateShadowMemory(Queue));
+
+    // Return LastEvent in QueueInfo
+    auto Context = getContext(Queue);
+    auto ContextInfo = getContextInfo(Context);
+    auto QueueInfo = ContextInfo->getQueueInfo(Queue);
+
+    std::scoped_lock<ur_mutex> Guard(QueueInfo->Mutex);
+    Event = QueueInfo->LastEvent;
+    QueueInfo->LastEvent = nullptr;
+
+    return true;
+}
+
+void SanitizerInterceptor::postLaunchKernel(ur_kernel_handle_t Kernel,
+                                            ur_queue_handle_t Queue,
+                                            ur_event_handle_t *Event) {
+    auto Program = getProgram(Kernel);
+    ur_event_handle_t ReadEvent{};
+
+    // If kernel has defined SPIR_DeviceSanitizerReportMem, then we try to read it
+    // to host, but it's okay that it isn't defined
```

Comment:
In which case is SPIR_DeviceSanitizerReportMem not defined?

The below code isn't handling the case when DeviceGlobalVariableRead fails even if SPIR_DeviceSanitizerReportMem exists.

Reply:
> In which case is SPIR_DeviceSanitizerReportMem not defined?

When these are no memory read/write in the kernel, for example, an empty kernel.

Reply:
I think a better approach is to query if SPIR_DeviceSanitizerReportMem exists and then read. This would cover all fail cases.

Reply:
There's no API to query if the device global variables exist.

Reply:
both OpenCL and LevelZero have API to get a pointer to a global variable. Do you mean unified runtime doesn't have this API? Is this a gap in unified runtime?
I don't think it is elegant to write to a global variable without querying if it exists.

Reply:
I filed a issue to track this.

https://github.com/oneapi-src/unified-runtime/issues/1214